store/cache.go

```
@@ -0,0 +1,169 @@
+func (s changesByReverseMtime) Len() int           { return len(s) }
+func (s changesByReverseMtime) Swap(i, j int)      { s[i], s[j] = s[j], s[i] }
+func (s changesByReverseMtime) Less(i, j int) bool { return s[i].ModTime().After(s[j].ModTime()) }
+
+// cacheManager implements a downloadCache via content based hard linking
+type CacheManager struct {
+	cacheDir string
+	maxItems int
+}
+
+// NewCacheManager returns a new CacheManager with the given cacheDir
+// and the given maximum amount of items. The idea behind it is the
+// following algorithm:
+//
+// 1. When starting a download, check if it exists in $cacheDir
+// 2. If found, update its mtime, hardlink into target location, and
+//    return success
+// 3. If not found, download the snap
+// 4. On success, hardlink into $cacheDir/<digest>
+// 5. If cache dir has more than maxItems entries, remove oldest mtimes
+//    until it has maxItems
+//
+// The caching part is done here, the downloading happens in the store.go
+// code.
+func NewCacheManager(cacheDir string, maxItems int) *CacheManager {
+	return &CacheManager{
+		cacheDir: cacheDir,
+		maxItems: maxItems,
+	}
+}
+
+// Get gets the given sha3 content and puts it into targetPath
+func (cm *CacheManager) Get(sha3_384, targetPath string) error {
+	if !cm.Lookup(sha3_384) {
+		return fmt.Errorf("cannot find %q in %q", sha3_384, cm.cacheDir)
+	}
+	if err := os.Link(cm.path(sha3_384), targetPath); err != nil {
+		return err
+	}
+	return os.Chtimes(targetPath, time.Now(), time.Now())
+}
+
+// Put adds a new file to the cache
+func (cm *CacheManager) Put(sourcePath string) error {
+	// happens on e.g. `snap download` which runs as the user
+	if !osutil.IsWritable(cm.cacheDir) {
+		return nil
+	}
+
+	if err := os.MkdirAll(cm.cacheDir, 0700); err != nil && !os.IsExist(err) {
+		return err
+	}
+
+	sha3_384, err := cm.digest(sourcePath)
+	if err != nil {
+		return err
+	}
+	if cm.Lookup(sha3_384) {
+		return os.Chtimes(cm.path(sha3_384), time.Now(), time.Now())
```

Same about times.

Also, this should probably check the error returned by Chtimes to make sure the file wasn't just removed. We can ignore the error in those cases.